### PR TITLE
go_1_10: init at 1.10 and set as default

### PR DESCRIPTION
**Pull request by Mic92**
Changes are minor from 1.9, so let's just set it as default straight.

changelog: https://golang.org/doc/go1.10

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by Mic92** [comment]
@GrahamcOfBorg build docker direnv

**Comment by Mic92** [comment]
cc @cstrahan @zimbatm

File: pkgs/development/compilers/go/remove-test-pie.patch
Comment (Mic92) [inline pkgs/development/compilers/go/remove-test-pie.patch]: oh looks like `gofmt` has added this. I will remove it after borg's first trial.

**Comment by adisbladis** [comment]
@Mic92 Do you think we should also update `buildGoPackage` to use 1.10?

**Comment by Mic92** [comment]
@adisbladis Yes. I forgot about that.

**Comment by adisbladis** [comment]
Fyi there is another open PR that does this https://github.com/NixOS/nixpkgs/pull/35311/

**Comment by Mic92** [comment]
@adisbladis oh, I have not checked for it.

**Comment by Mic92** [comment]
@GrahamcOfBorg build go

File: pkgs/development/compilers/go/1.10.nix
Comment (Mic92) [inline pkgs/development/compilers/go/1.10.nix]: @velovix are you ok with that?

**Comment by Mic92** [comment]
@GrahamcOfBorg build docker

**Review by zimbatm — APPROVED** [review]
looks good!

**Comment by zimbatm** [comment]
we should probably only keep up to 3 concurrent releases of compilers

**Comment by adisbladis** [comment]
@zimbatm We just dropped 1.6 and 1.7 earlier today (https://github.com/NixOS/nixpkgs/pull/35327) so once this is merged we have 1.8, 1.9 and 1.10.

I would like to drop 1.8 before the 18.03 release too (only the 2 latest releases are supported by upstream) but dropping 1.8 is currently blocked by `perkeep` which does not build with 1.9+.

**Comment by lfam** [comment]
Hi,

I noticed that Go 1.10 has a new "build caching" mechanism, briefly described at `go help cache`. I wonder if anything needs to be adjusted in how Nix builds Go software to take advantage of it?

In Guix, I found that the new Go cache validator was not choosing to re-use built Go objects but instead rebuilt all the dependencies from scratch during each package build.

File: pkgs/development/compilers/go/1.10.nix
Comment (velovix) [reply pkgs/development/compilers/go/1.10.nix]: Sure, that's fine

Comment (velovix) [reply pkgs/development/compilers/go/1.10.nix]: Do we know that wkennington isn't interested in being a maintainer anymore?

**Comment by velovix** [comment]
@lfam It seems that Go 1.10 puts cached build output in `~/.cache/go-build` as expected and does not rebuild the package I checked unless I clean the cache myself. I think it works as expected.

**Review by velovix — APPROVED** [review]
Looks good!

**Comment by Mic92** [comment]
@lfam there is a NixCon talk about incremental compilation https://www.youtube.com/watch?v=AED-dgCa9GU&index=8&list=PLgknCdxP89ReQzhfKwMYjLdwWsc7us8ns 
The only thing I have seen implemented so far is `ccacheStdenv`. For nixpkgs itself go projects are probably the smallest problem time-wise. Speeding up projects like chromium builds would be interesting though to cut down the critical chain.

**Comment by lfam** [comment]
@velovix Interesting, I wouldn't expect that to work in the package build environment, although I really don't have strong knowledge of Nix compared to Guix.

File: pkgs/development/compilers/go/1.10.nix
Comment (Mic92) [reply pkgs/development/compilers/go/1.10.nix]: @velovix he has its fork of nixpkgs, he is working: https://github.com/wkennington?tab=overview&from=2017-12-01&to=2017-12-31

**Comment by velovix** [comment]
@lfam I may have misunderstood your question. Are you asking if nixpkgs can take advantage of Go's caching while building Go applications?

**Comment by lfam** [comment]
@velovix Yes, that's what I'm wondering. We do it in Guix but it doesn't work anymore with Go 1.10 and I'm hoping to have a conversation with other people using Go with a functional package manager :) I know this is not the right forum for the discussion. Feel free to ping me on Freenode, user lfam.

**Pull request merged**
